# Praktikum 3: Pengolahan Data dengan NumPy dan Pandas
**Nama:** Gathan Hilabi
**NIM:** 60324059

### 1. Menyiapkan Notebook dan Mengimpor Library

In [1]:
import numpy as np
import pandas as pd

### 2. Dasar Array NumPy

In [2]:
# Membuat array NumPy dari sebuah list
nilai_ujian = np.array([78, 85, 90, 65, 72])

print("Rata-rata:", np.mean(nilai_ujian))
print("Maksimum:", np.max(nilai_ujian))
print("Minimum:", np.min(nilai_ujian))
print("Std Deviasi:", np.std(nilai_ujian))

Rata-rata: 78.0
Maksimum: 90
Minimum: 65
Std Deviasi: 8.921883209278185


### 3. Membaca Dataset dengan Pandas

In [3]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.head() # menampilkan 5 baris pertama

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


### 4. Eksplorasi Awal Dataset

In [4]:
print(df.shape) # jumlah baris & kolom
print(df.columns) # nama-nama kolom

df.info() # tipe data & jumlah data tidak kosong
df.describe() # statistik deskriptif kolom numerik

(891, 15)
Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    object 
 13  alive        891 non-null    object 
 14  alone        891 no

,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


### 5. Mengakses Kolom dan Baris

In [5]:
# Mengakses satu kolom
df["age"].head()

# Mengakses beberapa kolom sekaligus
df[["age", "sex", "survived"]].head()

# Memfilter baris berdasarkan kondisi
penumpang_anak = df[df["age"] < 18]
print("Jumlah penumpang anak:", len(penumpang_anak))

Jumlah penumpang anak: 113


### 6. Mendeteksi Missing Value

In [6]:
print(df.isnull().sum())

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64


### 7. Menangani Missing Value

In [7]:
# Strategi 1: isi nilai kosong dengan rata-rata (untuk data numerik)
df["age"] = df["age"].fillna(df["age"].mean())

# Strategi 2: hapus baris yang datanya kosong pada kolom tertentu
df = df.dropna(subset=["embarked"])

print(df.isnull().sum()) # periksa kembali

survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
embarked         0
class            0
who              0
adult_male       0
deck           688
embark_town      0
alive            0
alone            0
dtype: int64


### 8. Mendeteksi dan Menghapus Duplikasi

In [8]:
print("Jumlah data sebelum:", len(df))
print("Jumlah duplikasi:", df.duplicated().sum())

df = df.drop_duplicates()
print("Jumlah data sesudah:", len(df))

Jumlah data sebelum: 889
Jumlah duplikasi: 107
Jumlah data sesudah: 782


### 9. Menyimpan Dataset yang Sudah Dibersihkan

In [9]:
df.to_csv("titanic_bersih.csv", index=False)

### 10. Mendokumentasikan Proses dan Mengumpulkan Notebook
## Ringkasan Pembersihan Data
- Dataset awal berisi 891 baris dan 15 kolom.
- Kolom age memiliki 177 data kosong, diisi dengan nilai rata-rata.
- Baris dengan kolom embarked kosong dihapus.
- Ditemukan 107 baris duplikat dan telah dihapus.
- Dataset akhir berisi 782 baris siap dipakai untuk visualisasi.

---
## F. Latihan Mandiri

**Latihan 1:** Tampilkan 10 baris pertama dan 5 baris terakhir dataset menggunakan head(10) dan tail(5).

In [10]:
display(df.head(10))
display(df.tail(5))

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.000000,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.000000,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.000000,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.000000,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.000000,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
5,0,3,male,29.699118,0,0,8.4583,Q,Third,man,True,NaN,Queenstown,no,True
6,0,1,male,54.000000,0,0,51.8625,S,First,man,True,E,Southampton,no,True
7,0,3,male,2.000000,3,1,21.0750,S,Third,child,False,NaN,Southampton,no,False
8,1,3,female,27.000000,0,2,11.1333,S,Third,woman,False,NaN,Southampton,yes,False
9,1,2,female,14.000000,1,0,30.0708,C,Second,child,False,NaN,Cherbourg,yes,False


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
885,0,3,female,39.000000,0,5,29.125,Q,Third,woman,False,NaN,Queenstown,no,False
887,1,1,female,19.000000,0,0,30.000,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,29.699118,1,2,23.450,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.000000,0,0,30.000,C,First,man,True,C,Cherbourg,yes,True
890,0,3,male,32.000000,0,0,7.750,Q,Third,man,True,NaN,Queenstown,no,True


**Latihan 2:** Hitung rata-rata usia penumpang menggunakan NumPy (np.mean(df['age'])) dan Pandas (df['age'].mean()). Bandingkan apakah hasilnya sama.

In [11]:
mean_np = np.mean(df['age'])
mean_pd = df['age'].mean()
print("Rata-rata (NumPy):", mean_np)
print("Rata-rata (Pandas):", mean_pd)
print("Apakah sama?", mean_np == mean_pd)

Rata-rata (NumPy): 29.79479088310516
Rata-rata (Pandas): 29.79479088310516
Apakah sama? True


**Latihan 3:** Tampilkan data penumpang yang berada di kelas 1 (pclass == 1) dan selamat (survived == 1) menggunakan filter kondisi ganda (gunakan operator &).

In [12]:
penumpang_selamat_kelas1 = df[(df['pclass'] == 1) & (df['survived'] == 1)]
display(penumpang_selamat_kelas1.head())

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
1,1,1,female,38.000000,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
3,1,1,female,35.000000,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
11,1,1,female,58.000000,0,0,26.5500,S,First,woman,False,C,Southampton,yes,True
23,1,1,male,28.000000,0,0,35.5000,S,First,man,True,A,Southampton,yes,True
31,1,1,female,29.699118,1,0,146.5208,C,First,woman,False,B,Cherbourg,yes,False


**Latihan 4:** Isi missing value pada kolom embarked dengan nilai yang paling sering muncul (modus), menggunakan df['embarked'].mode()[0].

In [13]:
# Catatan: karena baris dengan embarked kosong sudah didrop sebelumnya, baris ini sebagai latihan logika saja.
df['embarked'] = df['embarked'].fillna(df['embarked'].mode()[0])
print(df.isnull().sum())

survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
embarked         0
class            0
who              0
adult_male       0
deck           582
embark_town      0
alive            0
alone            0
dtype: int64


**Latihan 5:** Hitung berapa persentase penumpang yang selamat dari keseluruhan data, menggunakan df['survived'].mean() * 100.

In [14]:
persentase_selamat = df['survived'].mean() * 100
print(f"Persentase penumpang selamat: {persentase_selamat:.2f}%")

Persentase penumpang selamat: 41.05%


---
## H. Pertanyaan Diskusi

**1. Mengapa penting memeriksa dan menangani missing value sebelum data digunakan untuk membuat visualisasi?**
Karena data yang hilang dapat menyebabkan kesalahan dalam analisis atau visualisasi, mengganggu perhitungan statistik, atau menyebabkan fungsi/library plotting error. Menangani missing value memastikan visualisasi merepresentasikan data dengan akurat tanpa adanya bias atau kegagalan komputasi.

**2. Dalam kondisi seperti apa sebaiknya kita menggunakan dropna(), dan dalam kondisi seperti apa lebih baik menggunakan fillna()?**
- Gunakan `dropna()` saat jumlah missing value sangat sedikit dibandingkan total data, atau jika nilai tersebut krusial (misal label target) sehingga menebak nilainya berisiko menghasilkan kesimpulan yang salah.
- Gunakan `fillna()` saat kita masih ingin mempertahankan data tersebut (karena fiturnya penting atau jumlah observasi sedikit), dan kita bisa menggunakan pendekatan statistik (seperti rata-rata, median, modus) untuk mengisi kekosongan tanpa banyak mengubah distribusi data.

**3. Apa risiko yang bisa terjadi jika data duplikat tidak dihapus sebelum data dianalisis atau divisualisasikan?**
Data duplikat dapat memberikan bobot berlebih pada observasi tertentu, mendistorsi hasil analisis, statistik deskriptif, dan juga bentuk visualisasi (misalnya histogram akan memiliki frekuensi yang berlebihan di titik tertentu), sehingga kesimpulan yang ditarik menjadi bias dan kurang akurat.